certbot/compat/os.py

```diff
@@ -0,0 +1,154 @@
+"""
+This compat modules extends native capabilities of core os module to handle correctly platform
+specific operations (eg. chown, chmod, geuid).
+This module is intended to replace standard os module throughout certbot projects (except acme)
+"""
+from __future__ import absolute_import
+
+# Expose everything from standard os package to make current package a complete replacement of os.
+# pylint: disable=wildcard-import,unused-wildcard-import,redefined-builtin
+from os import *  # type: ignore
+# pylint: enable=wildcard-import,unused-wildcard-import,redefined-builtin
+
+import errno
+import os as std_os
+import sys
+
+from acme.magic_typing import Callable, Union  # pylint: disable=unused-import, no-name-in-module
+from certbot.compat import security
+
+# Monkey patch ourselves to get os attributes that are not in __all__ (so not from os import *).
+ourselves = sys.modules[__name__]
```

Comment:
why not just install ourselves in sys.modules as `os`? then we don't have to replace the call everywhere, and don't need to add the lint changes.

Reply:
Well, it is a really good question. It was the first thing in my mind: reinjecting a corrected os module in the modules pool of current Python instance. Then it would be used seamlessly by any Certbot module.

There is a lot of problems with that approach eventually.

First, the corrected os module would be used not only by certbot modules, but also by every dependency of Certbot. As you can see, the functionalities added are very specific to Certbot needs. Because in fact, Unix and Windows security models are just too different to have a global and bijective representation in both platforms (otherwise it would be implemented in the standard library). So it means that this model could not be suited for the dependencies, and add unexpected behavior to them. This security model, and so the associated modules, should only be used for the sole Certbot purpose.

Second, the os module is swapped only when the relevant code to manipulate sys.modules is called. Obviously, it should be added to the `__init__.py` of `certbot` package. But also to any other package, like the ones from dns plugins and authenticators, to avoid to break any permission. And ensure this, is the same problem as ensuring that standard `os` is not usedm and should be corrected by a lint process. So lint rules may or may not be used to enforce this policy, but using an injection in sys.modules, do not remove on its own the requirement to check the code.

Finally, even if implicit approach fit quite well in Python, I think there is a limit, related to good or bad design decisions. Making implicit modifications of one of the most used standard library feels wrong to me. I largely prefer to be explicit about using a different `os` module, and make it clear from the code.

Reply:
> First, the corrected os module would be used not only by certbot modules, but also by every dependency of Certbot. As you can see, the functionalities added are very specific to Certbot needs. Because in fact, Unix and Windows security models are just too different to have a global and bijective representation in both platforms (otherwise it would be implemented in the standard library). So it means that this model could not be suited for the dependencies, and add unexpected behavior to them. This security model, and so the associated modules, should only be used for the sole Certbot purpose.

Is it ever the case that for any of our dependencies that create files, we would want files to be created correctly in Certbot but incorrectly in their own code?

> Second, the os module is swapped only when the relevant code to manipulate sys.modules is called. Obviously, it should be added to the __init__.py of certbot package. But also to any other package, like the ones from dns plugins and authenticators, to avoid to break any permission. And ensure this, is the same problem as ensuring that standard os is not usedm and should be corrected by a lint process. So lint rules may or may not be used to enforce this policy, but using an injection in sys.modules, do not remove on its own the requirement to check the code.

If it's true that our dependencies get the patched version of `os`, then our plugins will also get it and won't need to patch it themselves. This keeps the changed code to the smallest possible footprint, which is good for maintenance (if we need to change it in the future) and for packaging (if a unix packager doesn't want the windows support code in there at all).

> Finally, even if implicit approach fit quite well in Python, I think there is a limit, related to good or bad design decisions. Making implicit modifications of one of the most used standard library feels wrong to me. I largely prefer to be explicit about using a different os module, and make it clear from the code.

This isn't something that callers of `os` within Certbot should have to worry about. We should expect that people using the functions shouldn't have to understand how they work to be able to call them, and leave the behavior similar enough that it works as expected. Requiring callers to import a different class will just lead to changing the line, not using the functions any differently, which is just an implementation blocker.

Reply:
OK I took the wrong examples with plugins, forgotting that plugins do not call certbot, is it certbot that call plugins through the dynamic plugins discovery mechanism of python.

That was not the point I wanted to enlight. 

Monkeypatching a standard library is not a question about design, it is a question about wrong code. If you monkeypatch the standard library, it does not make it effective only for your own code, but for any code that use yours.

It means that any external developper, for any project and any context, import certbot package in one of its own module, will see a standard module be silently swapped to a custom on, starting from the first import of the certbot package, and for the rest of the runtime execution.

The consequences are widely unknown, completely out of our range. It is just not an option. We can discuss ways to ease usage of this specific module, but monkeypatching is not a solution.

Apart that, developers are not required to understand the compat.os module. It is an implementation detail that claim externally as a contract to be a valid replacement of the standard os module for certbot. I do not think that it is overcomitting, as it does not add a single line compared to a call to the standard os, and that a CI process is able to do the job to check this specific subtility. Compared with the thousands of unit test out there, and all the dynamic mocks called from them, for me it is not a big deal ^^